### ISSUE TITLE
okay-eval-deepseq

### ISSUE THREAD

**Issue by jwiegley** [haskell-nix/hnix#141]
@ryantrinkle Do you think this should evaluate to 456:

```
builtins.deepSeq (let as = { x = 123; y = as; }; in as) 456
```

If so, exactly what are the semantics of `deepSeq`? It says it evaluates recursively, which means it should evaluate every `y.y.y.....` at all depths, leading to memory exhaustion.

The current version of hnix detects the evaluation cycle now.

**Comment by ryantrinkle** [comment]
Yep, per nix, that evaluates to 456.  Gross, if you ask me.

```
nix-repl> builtins.deepSeq (let as = { x = 123; y = as; }; in as) 456
456
```

**Comment by jwiegley** [comment]
What I wonder is: how can that evaluate to 456? It's as if they stop evaluating once they reach the cycle, but that's not the meaning of "deep seq".

**Comment by ryantrinkle** [comment]
I think they treat data structure cycles different from cyclic computations.  I'm not sure what (if any) underlying theory would describe it, though.

**Comment by jwiegley** [comment]
Then I may need to add some kind of coloring to `normalForm` to know when to stop recursing, since I don't think we can detect this generally during evaluation and still allow the `functionargs` example to evaluate.

**Comment by jwiegley** [comment]
Fixed by 8e5141fec5478e5538b9b938085585eab7b7ebf9
